File: glox-treewalk/internal/syntax/scanner.go
package syntax

import (
	"fmt"
	"strconv"

	"github.com/littlekuo/glox-treewalk/internal/util"
)

var keywords = map[string]TokenType{
	"and":    TOKEN_AND,
	"class":  TOKEN_CLASS,
	"else":   TOKEN_ELSE,
	"false":  TOKEN_FALSE,
	"for":    TOKEN_FOR,
	"fun":    TOKEN_FUN,
	"if":     TOKEN_IF,
	"nil":    TOKEN_NIL,
	"or":     TOKEN_OR,
	"print":  TOKEN_PRINT,
	"return": TOKEN_RETURN,
	"super":  TOKEN_SUPER,
	"this":   TOKEN_THIS,
	"true":   TOKEN_TRUE,
	"var":    TOKEN_VAR,
	"while":  TOKEN_WHILE,
	"break":  TOKEN_BREAK,
}

type Scanner struct {
	source  string
	tokens  []Token
	start   int
	current int
	// line number
	line    int
	scanErr error
}

func NewScanner(source string) *Scanner {
	return &Scanner{
		source: source,
		tokens: make([]Token, 0),
		line:   1,
	}
}

func (s *Scanner) ScanTokens() []Token {
	for {
		if s.isEnd() {
			break
		}
		s.start = s.current
		s.scanToken()
	}

	// the last token
	s.tokens = append(s.tokens, NewToken(TOKEN_EOF, "", nil, s.line))
	return s.tokens
}

func (s *Scanner) scanToken() {
	c := s.advance()
	switch c {
	case '(':
		s.addSimpleToken(TOKEN_LEFT_PAREN)
	case ')':
		s.addSimpleToken(TOKEN_RIGHT_PAREN)
	case '{':
		s.addSimpleToken(TOKEN_LEFT_BRACE)
	case '}':
		s.addSimpleToken(TOKEN_RIGHT_BRACE)
	case ',':
		s.addSimpleToken(TOKEN_COMMA)
	case '.':
		s.addSimpleToken(TOKEN_DOT)
	case '-':
		s.addSimpleToken(TOKEN_MINUS)
	case '+':
		s.addSimpleToken(TOKEN_PLUS)
	case ';':
		s.addSimpleToken(TOKEN_SEMICOLON)
	case '*':
		s.addSimpleToken(TOKEN_STAR)
	case '!':
		s.addConditionalToken('=', TOKEN_BANG_EQUAL, TOKEN_BANG)
	case '=':
		s.addConditionalToken('=', TOKEN_EQUAL_EQUAL, TOKEN_EQUAL)
	case '<':
		s.addConditionalToken('=', TOKEN_LESS_EQUAL, TOKEN_LESS)
	case '>':
		s.addConditionalToken('=', TOKEN_GREATER_EQUAL, TOKEN_GREATER)
	case '/':
		if s.match('/') {
			// A comment goes until the end of the line.
			for s.peek() != '\n' && !s.isEnd() {
				s.advance()
			}
		} else if s.match('*') {
			s.scanBlockComment()
		} else {
			s.addSimpleToken(TOKEN_SLASH)
		}
	case ' ', '\r', '\t':
		// Ignore whitespace.
	case '\n':
		s.line++
	case '"':
		s.scanString()
	default:
		if isDigit(c) {
			s.scanNumber()
		} else if isAlpha(c) {
			s.scanIdentifier()
		} else {
			s.error(s.line, "Unexpected character.")
		}
	}
}

func (s *Scanner) addSimpleToken(tk TokenType) {
	s.addTokenWithLiteral(tk, nil)
}

func (s *Scanner) addTokenWithLiteral(tk TokenType, literal any) {
	text := s.source[s.start:s.current]
	s.tokens = append(s.tokens, NewToken(tk, text, literal, s.line))
}

func (s *Scanner) isEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) GetError() error {
	return s.scanErr
}

func (s *Scanner) addConditionalToken(expected byte, matchedType TokenType, unmatchedType TokenType) {
	if s.match(expected) {
		s.addSimpleToken(matchedType)
	} else {
		s.addSimpleToken(unmatchedType)
	}
}

func (s *Scanner) match(expected byte) bool {
	if s.isEnd() {
		return false
	}
	if s.source[s.current] != expected {
		return false
	}

	// if match, then advance
	s.current++
	return true
}

func (s *Scanner) advance() byte {
	s.current++
	return s.source[s.current-1]
}

func (s *Scanner) peek() byte {
	if s.isEnd() {
		return 0
	}
	return s.source[s.current]
}

func (s *Scanner) peekNext() byte {
	if s.current+1 >= len(s.source) {
		return 0
	}
	return s.source[s.current+1]
}

func (s *Scanner) scanString() {
	for s.peek() != '"' && !s.isEnd() {
		if s.peek() == '\n' {
			s.line++
		}
		s.advance()
	}

	if s.isEnd() {
		s.error(s.line, "Unterminated string")
		return
	}

	// the closing ".
	s.advance()

	s.addTokenWithLiteral(TOKEN_STRING, s.source[s.start+1:s.current-1])
}

// support: 1234, 12.34
// not support: 1234. , .1234
func (s *Scanner) scanNumber() {
	for isDigit(s.peek()) {
		s.advance()
	}
	if s.peek() == '.' && isDigit(s.peekNext()) {
		// Consume the "."
		s.advance()

		for isDigit(s.peek()) {
			s.advance()
		}
	}

	floatValue, _ := strconv.ParseFloat(s.source[s.start:s.current], 64)
	s.addTokenWithLiteral(TOKEN_NUMBER, floatValue)
}

func (s *Scanner) scanIdentifier() {
	for isAlphaNumeric(s.peek()) {
		s.advance()
	}

	text := s.source[s.start:s.current]
	tokenType, ok := keywords[text]
	if ok {
		s.addSimpleToken(tokenType)
	} else {
		s.addSimpleToken(TOKEN_IDENTIFIER)
	}
}

// support: /* */
func (s *Scanner) scanBlockComment() {
	nestingLevel := 1 // 初始嵌套层级
	for nestingLevel > 0 && !s.isEnd() {
		switch {
		case s.peek() == '\n':
			s.line++
			s.advance()
		case s.peek() == '/' && s.peekNext() == '*':
			s.advance()
			s.advance()
			nestingLevel++
		case s.peek() == '*' && s.peekNext() == '/':
			s.advance()
			s.advance()
			nestingLevel--
		default:
			s.advance()
		}
	}

	if nestingLevel > 0 {
		s.error(s.line, "unterminated block comment")
	}
}

func (s *Scanner) error(line int, message string) {
	s.scanErr = util.ErrorMsg(line, message)
	fmt.Printf("scann error at line %d: %s\n", line, message)
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
}

func isAlphaNumeric(c byte) bool {
	return isAlpha(c) || isDigit(c)
}

File: llox/src/parser/scanner.cpp
#include "parser/scanner.h"
#include <any>
#include <cctype>
#include <cmath>
#include <iostream>

const std::unordered_map<std::string, TokenType> keywords_ = {
    {"and", TokenType::TOKEN_AND},
    {"class", TokenType::TOKEN_CLASS},
    {"else", TokenType::TOKEN_ELSE},
    {"false", TokenType::TOKEN_FALSE},
    {"for", TokenType::TOKEN_FOR},
    {"fun", TokenType::TOKEN_FUN},
    {"if", TokenType::TOKEN_IF},
    {"nil", TokenType::TOKEN_NIL},
    {"or", TokenType::TOKEN_OR},
    {"print", TokenType::TOKEN_PRINT},
    {"return", TokenType::TOKEN_RETURN},
    {"super", TokenType::TOKEN_SUPER},
    {"this", TokenType::TOKEN_THIS},
    {"true", TokenType::TOKEN_TRUE},
    {"var", TokenType::TOKEN_VAR},
    {"while", TokenType::TOKEN_WHILE},
    {"break", TokenType::TOKEN_BREAK},
    {"continue", TokenType::TOKEN_CONTINUE}};

Scanner::Scanner(const std::string &source) : source_(source) {}

void Scanner::scan_tokens(std::vector<Token> &tokens) {
  while (!is_end()) {
    start_ = current_;
    start_column_ = current_column_;
    start_line_ = current_line_;
    scan_token(tokens);
  }
  tokens.emplace_back(TokenType::TOKEN_EOF, "", 0, start_line_, start_column_);
}

bool Scanner::has_error() const { return has_error_; }

void Scanner::scan_token(std::vector<Token> &tokens) {
  char c = advance();
  switch (c) {
  case '(':
    add_token(tokens, TokenType::TOKEN_LEFT_PAREN);
    break;
  case ')':
    add_token(tokens, TokenType::TOKEN_RIGHT_PAREN);
    break;
  case '{':
    add_token(tokens, TokenType::TOKEN_LEFT_BRACE);
    break;
  case '}':
    add_token(tokens, TokenType::TOKEN_RIGHT_BRACE);
    break;
  case ',':
    add_token(tokens, TokenType::TOKEN_COMMA);
    break;
  case '.':
    add_token(tokens, TokenType::TOKEN_DOT);
    break;
  case '-':
    add_token(tokens, TokenType::TOKEN_MINUS);
    break;
  case '+':
    add_token(tokens, TokenType::TOKEN_PLUS);
    break;
  case ';':
    add_token(tokens, TokenType::TOKEN_SEMICOLON);
    break;
  case '*':
    add_token(tokens, TokenType::TOKEN_STAR);
    break;

  case '!':
    add_conditional_token(tokens, '=', TokenType::TOKEN_BANG_EQUAL,
                          TokenType::TOKEN_BANG);
    break;
  case '=':
    add_conditional_token(tokens, '=', TokenType::TOKEN_EQUAL_EQUAL,
                          TokenType::TOKEN_EQUAL);
    break;
  case '<':
    add_conditional_token(tokens, '=', TokenType::TOKEN_LESS_EQUAL,
                          TokenType::TOKEN_LESS);
    break;
  case '>':
    add_conditional_token(tokens, '=', TokenType::TOKEN_GREATER_EQUAL,
                          TokenType::TOKEN_GREATER);
    break;

  case '/':
    if (match('/')) { // handle single-line comment
      while (peek() != '\n' && !is_end())
        advance();
    } else if (match('*')) { // handle multi-line comment
      scan_block_comment();
    } else {
      add_token(tokens, TokenType::TOKEN_SLASH);
    }
    break;

  case ' ':
  case '\r':
  case '\t':
    break;

  case '\n':
    break;

  case '"':
    scan_string(tokens);
    break;

  default:
    if (isdigit(c)) {
      scan_number(tokens);
    } else if (isalpha(c) || c == '_') {
      scan_identifier(tokens);
    } else {
      error("Unexpected character");
    }
  }
}

bool Scanner::is_end() const { return current_ >= source_.size(); }

bool Scanner::match(char expected) {
  if (is_end())
    return false;
  if (source_[current_] != expected)
    return false;
  current_++;
  return true;
}

char Scanner::advance() {
  char c = source_[current_];
  current_++;
  if (c == '\n') {
    current_line_++;
    current_column_ = 0;
  } else {
    current_column_++;
  }
  return c;
}

char Scanner::peek() const {
  if (is_end())
    return '\0';
  return source_[current_];
}

char Scanner::peek_next() const {
  if (current_ + 1 >= source_.size())
    return '\0';
  return source_[current_ + 1];
}

void Scanner::add_token(std::vector<Token> &tokens, TokenType type) {
  tokens.emplace_back(type, source_.c_str() + start_, current_ - start_,
                      start_line_, start_column_);
}

void Scanner::add_token(std::vector<Token> &tokens, TokenType type,
                        const char *start, int len) {
  tokens.emplace_back(type, start, len, start_line_, start_column_);
}

void Scanner::add_conditional_token(std::vector<Token> &tokens, char expected,
                                    TokenType matched, TokenType unmatched) {
  if (match(expected)) {
    add_token(tokens, matched);
  } else {
    add_token(tokens, unmatched);
  }
}

void Scanner::error(const std::string &message) {
  std::cout << "error at line " << current_line_ << ": " << message
            << std::endl;
  has_error_ = true;
}

void Scanner::scan_string(std::vector<Token> &tokens) {
  while (peek() != '"' && !is_end()) {
    advance();
  }
  if (is_end()) {
    error("Unterminated string");
    return;
  }
  advance(); // consume the closing quote
  add_token(tokens, TokenType::TOKEN_STRING, source_.c_str() + start_ + 1,
            current_ - start_ - 2);
}

void Scanner::scan_number(std::vector<Token> &tokens) {
  while (isdigit(peek()))
    advance();
  if (peek() == '.' && isdigit(peek_next())) {
    advance();
    while (isdigit(peek()))
      advance();
  }
  add_token(tokens, TokenType::TOKEN_NUMBER, source_.c_str() + start_,
            current_ - start_);
}

void Scanner::scan_identifier(std::vector<Token> &tokens) {
  while (isalnum(peek()) || peek() == '_')
    advance();
  std::string text = source_.substr(start_, current_ - start_);
  TokenType type = keywords_.count(text) > 0 ? keywords_.at(text)
                                             : TokenType::TOKEN_IDENTIFIER;
  add_token(tokens, type);
}

void Scanner::scan_block_comment() {
  int depth = 1;
  while (depth > 0 && !is_end()) {
    if (peek() == '/' && peek_next() == '*') {
      advance();
      advance();
      depth++;
    } else if (peek() == '*' && peek_next() == '/') {
      advance();
      advance();
      depth--;
    } else {
      advance();
    }
  }
  if (depth > 0) {
    error("Unterminated block comment");
  }
}